sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java

```diff
@@ -0,0 +1,250 @@
+        policies.addAll(perCallPolicies);
+        HttpPolicyProviders.addBeforeRetryPolicies(policies);
+
+        policies.add(retryPolicy == null ? new RetryPolicy() : retryPolicy);
+        policies.add(new CookiePolicy());
+        policies.add(new AddDatePolicy());
+
+        policies.addAll(perRetryPolicies);
+        HttpPolicyProviders.addAfterRetryPolicies(policies);
+        HttpLoggingPolicy loggingPolicy = new HttpLoggingPolicy(logOptions);
+
+        // We generally put credential policy between BeforeRetry and AfterRetry policies and put Logging policy in the end.
+        // However since ACR uses the rest endpoints of the service in the credential policy,
+        // we want to be able to use the same pipeline (minus the credential policy) to have uniformity in the policy
+        // pipelines across all ACR endpoints.
+        if (credential == null) {
+            logger.verbose("Credentials are null, enabling anonymous access");
+        }
+
+        ArrayList<HttpPipelinePolicy> credentialPolicies = clone(policies);
+        credentialPolicies.add(loggingPolicy);
+
+        ContainerRegistryTokenService tokenService = new ContainerRegistryTokenService(
+            credential,
+            audience,
+            endpoint,
+            serviceVersion,
+            new HttpPipelineBuilder()
+                .policies(credentialPolicies.toArray(new HttpPipelinePolicy[0]))
+                .httpClient(httpClient)
+                .build(),
+            JacksonAdapter.createDefaultSerializerAdapter());
+
+        ContainerRegistryCredentialsPolicy credentialsPolicy = new ContainerRegistryCredentialsPolicy(tokenService);
+        policies.add(credentialsPolicy);
+
+        policies.add(loggingPolicy);
+        HttpPipeline httpPipeline =
+            new HttpPipelineBuilder()
+                .policies(policies.toArray(new HttpPipelinePolicy[0]))
+                .httpClient(httpClient)
+                .build();
+        return httpPipeline;
+    }
+
+    private static ArrayList<HttpPipelinePolicy> clone(ArrayList<HttpPipelinePolicy> policies) {
+        ArrayList<HttpPipelinePolicy> clonedPolicy = new ArrayList<>();
+        for (HttpPipelinePolicy policy:policies) {
+            clonedPolicy.add(policy);
+        }
+
+        return clonedPolicy;
+    }
+
+    /**
+     * This method computes the digest for the buffer content.
+     * Docker digest is a SHA256 hash of the docker image content and is deterministic based on the image build.
+     * @param buffer The buffer containing the image bytes.
+     * @return SHA-256 digest for the given buffer.
+     */
+    public static String computeDigest(ByteBuffer buffer) {
+        int startPosition = buffer.position();
+        try {
+            MessageDigest md = MessageDigest.getInstance("SHA-256");
+            md.update(buffer);
+            byte[] digest = md.digest();
+            return "sha256:" + byteArrayToHex(digest);
+
+        } catch (NoSuchAlgorithmException e) {
```

Comment:
Should this exception really be swallowed?

Reply:
No, I have created an issue to track fixing this part. [https://github.com/Azure/azure-sdk-for-java/issues/27088](https://github.com/Azure/azure-sdk-for-java/issues/27088)